src/test/regress/expected/merge.out

```diff
@@ -1213,9 +1315,1231 @@ SELECT * FROM ft_target;
+                     ->  Sort  (cost=83.37..86.37 rows=1200 width=10)
+                           Sort Key: sda.id
+                           ->  Seq Scan on target_json_4000094 sda  (cost=0.00..22.00 rows=1200 width=10)
+                     ->  Sort  (cost=83.37..86.37 rows=1200 width=4)
+                           Sort Key: sdn.id
+                           ->  Seq Scan on source_json_4000098 sdn  (cost=0.00..22.00 rows=1200 width=4)
+(47 rows)
+
+CREATE FUNCTION immutable_hash(int) RETURNS int
+AS 'SELECT hashtext( ($1 + $1)::text);'
+LANGUAGE SQL
+IMMUTABLE
+RETURNS NULL ON NULL INPUT;
+MERGE INTO target_json sda
+USING source_json sdn
+ON sda.id = sdn.id
+WHEN matched THEN
+	UPDATE SET z = immutable_hash(sdn.z);
+-- Test bigserial
+CREATE TABLE source_serial (id integer, z int, d bigserial);
+CREATE TABLE target_serial (id integer, z int, d bigserial);
+INSERT INTO source_serial SELECT i,i FROM generate_series(0,100)i;
+SELECT create_distributed_table('source_serial', 'id'), create_distributed_table('target_serial', 'id');
+NOTICE:  Copying data from local table...
+NOTICE:  copying the data has completed
+DETAIL:  The local data in the table is no longer visible, but is still on disk.
+HINT:  To remove the local data, run: SELECT truncate_local_data_after_distributing_table($$merge_schema.source_serial$$)
+ create_distributed_table | create_distributed_table
+---------------------------------------------------------------------
+                          |
+(1 row)
+
+MERGE INTO target_serial sda
+USING source_serial sdn
+ON sda.id = sdn.id
+WHEN NOT matched THEN
+       INSERT (id, z) VALUES (id, z);
+SELECT count(*) from source_serial;
+ count
+---------------------------------------------------------------------
+   101
+(1 row)
+
+SELECT count(*) from target_serial;
+ count
+---------------------------------------------------------------------
+   101
+(1 row)
+
 --
 -- Error and Unsupported scenarios
 --
+-- zero shard query as source_json is zero shard
+EXPLAIN MERGE INTO target_json sda
+USING (SELECT * FROM source_json WHERE false) sdn
+ON sda.id = sdn.id AND sda.id = 2
+WHEN NOT matched THEN
+	INSERT (id, z) VALUES (sdn.id, 5);
+ERROR:  MERGE command is only supported when distributed tables are joined on their distribution column
+-- modifying CTE not supported
+EXPLAIN
+WITH cte_1 AS (DELETE FROM target_json)
+MERGE INTO target_json sda
+USING source_json sdn
+ON sda.id = sdn.id
+WHEN NOT matched THEN
+	INSERT (id, z) VALUES (sdn.id, 5);
+ERROR:  MERGE command is only supported when distributed tables are joined on their distribution column
+-- Grouping sets not supported
+MERGE INTO citus_target t
+USING (SELECT count(*), id FROM citus_source GROUP BY GROUPING SETS (id, val)) subq
+ON subq.id = t.id
+WHEN MATCHED AND t.id > 350 THEN
+    UPDATE SET val = t.val || 'Updated'
+WHEN NOT MATCHED THEN
+        INSERT VALUES (subq.id, 99)
+WHEN MATCHED AND t.id < 350 THEN
+        DELETE;
+ERROR:  cannot push down this subquery
```

Comment:
some of the error messages are off:
```SQL
MERGE INTO prept                                                                                                     USING (SELECT 2 as s3, preps.* FROM (SELECT * FROM preps LIMIT 1) as foo LEFT JOIN preps USING( s1)) as foo                                                    ON prept.t1 = foo.s1                                                                                                                                           WHEN MATCHED THEN                                                                                                                                                      UPDATE SET t2 = t2 + 1                                                                                                                                 WHEN NOT MATCHED THEN                                                                                                                                                  INSERT VALUES(s1, s3);
ERROR:  cannot push down this subquery
DETAIL:  Limit clause is currently unsupported when a subquery references a column from another query
```

That is because the error messages assumes recursive planning kicks in on the subquery. I think we can ignore for now, just noting for completeness

Reply:
Added it as a negative test and also updated the PR description.